components/datetime/src/lib.rs

```diff
@@ -22,11 +22,9 @@
 //! let provider = icu_testdata::get_provider();
 //!
 //! // See the next code example for a more ergonomic example with .into().
-//! let options = DateTimeFormatOptions::Length(length::Bag {
-//!     date: Some(length::Date::Medium),
-//!     time: Some(length::Time::Short),
-//!     ..Default::default()
-//! });
+//! let options = DateTimeFormatOptions::Length(length::Bag::new(
+//!     Some(length::Date::Medium), Some(length::Time::Short),
+//! ));
```

Comment:
Without adding `::new` to the `length::Bag`, I think this can be

```rust
let mut options = length::Bag::default();
options.date = length::Date::Medium;
options.time = length::Time::Short;

let dtf = DateTimeFormat::<Gregorian>::try_new(locale!("en"), &provider, &options.into())
```

via an `impl From<length::Bag> for DateTimeFormatOptions` if such a thing doesn't exist yet

Reply:
Also, I think I will be advocating for splitting this into separate constructors anyway for code slicing / data slicing purposes. We will want `try_new_length()` and `try_new_components()` or something like that. We could do that now if you like.

Reply:
I'd rather not introduce even more changes to this PR.

I find the `::new()` thing to be easier to work with than the mutable piece-by-piece thing, I added it on purpose to avoid having to do it this way instead.


I feel like further stuff can be done by component owners in a followup.

Reply:
Could you instead introduce `From<(Option<DateStyle>, Option<TimeStyle>)> for DateTimeFormatOptions`?

I feel more comfortable with stability of such interface.

Reply:
That's extremely unidiomatic.

I think `new()` would be stable, if we add more things we can add `new_with_foo()` `new_with_bar()` etc

Reply:
> That's extremely unidiomatic.

Why is it unidiomatic? We have so many cases where we have `X` and `Y` and we often want to compose `Z` that uses both `X` and `Y` that we introduce `From<(X, Y)> for Z`. It seems safe because the odds that `Z` will ever *not* have `X` and `Y` are very low, and in every other case it still allows for other `From`'s and other constructors.

> I think new() would be stable, if we add more things we can add new_with_foo() new_with_bar() etc

I disagree with this assessment. I think we're yet to cool down our exploration of options for DTF and what is `new` vs `new_x` shouldn't be decided yet.

Reply:
@zbraniecki note that when I say "stable" here, I mean "stable post 1.0". Can leave a comment to that effect.

> Why is it unidiomatic?

Things like `MarkerForDifferentFromImpl(foo).into()` or `(foo, bar).into()` is about constructing a specific type to direct `From` dispatch and in that case it's always considered better to have a method, since you can just _name_ the method.

The purpose of `From` is to give you convenient ways to convert between types you have, not a separate layer of method dispatch where you don't have to name things. 

I'm happy to name it `new_for_date_time()` instead.

Reply:
> I'm happy to name it new_for_date_time() instead.

Can we use `from_date_time_style` ?